android/src/main/java/com/marigold/rnsdk/RNMessageStreamModule.kt

```diff
@@ -27,20 +31,45 @@ class RNMessageStreamModule (reactContext: ReactApplicationContext, private val
     @VisibleForTesting
     internal var jsonConverter: JsonConverter = JsonConverter()
 
+    private val eventListener = EventListener()
+    private var showDefaultInAppNotification = true
+
     init {
         messageStream.setOnInAppNotificationDisplayListener(this)
     }
 
     override fun shouldPresentInAppNotification(message: Message): Boolean {
-        try {
-            val writableMap = jsonConverter.convertJsonToMap(message.toJSON())
-            reactApplicationContext
+        return runBlocking {
+            val result = async { emitWithTimeout(message) }
+            result.await()
+        }
+    }
+
+    private suspend fun emitWithTimeout(message: Message): Boolean {
+        return withTimeoutOrNull(5000L) {
+            try {
+                val writableMap = jsonConverter.convertJsonToMap(message.toJSON())
+                val emitter = reactApplicationContext
                     .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
-                    .emit("inappnotification", writableMap)
-        } catch (e: JSONException) {
-            e.printStackTrace()
+                emitter.emit("inappnotification", writableMap)
+                waitForAcknowledgment()
+            } catch (e: JSONException) {
+                e.printStackTrace()
+                showDefaultInAppNotification
+            }
+        } ?: showDefaultInAppNotification
+    }
+
+    private suspend fun waitForAcknowledgment(): Boolean {
+        while (showDefaultInAppNotification) {
```

Comment:
The value for `showDefaultInAppNotification` would not be reset after the first call, so this delay would be skipped for every in-app after the first one. Maybe take a look at `Channel<Boolean>`? (https://kotlinlang.org/docs/channels.html) You could block waiting on the channel, then pass back a Boolean value in `acknowledgeEvent` which could be passed through to unblock it.